OpenRA.Mods.Common/Traits/TerrainLighting.cs

```diff
@@ -0,0 +1,140 @@
+	{
+		public readonly float Intensity = 1;
+		public readonly float HeightStep = 0;
+		public readonly float RedTint = 1;
+		public readonly float GreenTint = 1;
+		public readonly float BlueTint = 1;
+
+		[Desc("Size of light source partition bins (cells)")]
+		public readonly int BinSize = 10;
+
+		public override object Create(ActorInitializer init) { return new TerrainLighting(init.World, this); }
+	}
+
+	public sealed class TerrainLighting : ITerrainLighting
+	{
+		class LightSource
+		{
+			public readonly WPos Pos;
+			public readonly CPos Cell;
+			public readonly WDist Range;
+			public readonly float Intensity;
+			public readonly float3 Tint;
+
+			public LightSource(WPos pos, CPos cell, WDist range, float intensity, float3 tint)
+			{
+				Pos = pos;
+				Cell = cell;
+				Range = range;
+				Intensity = intensity;
+				Tint = tint;
+			}
+		}
+
+		readonly TerrainLightingInfo info;
+		readonly Map map;
+		readonly Dictionary<int, LightSource> lightSources = new Dictionary<int, LightSource>();
+		readonly SpatiallyPartitioned<LightSource> partitionedLightSources;
+		readonly float3 globalTint;
+		int nextLightSourceToken = 1;
+
+		public event Action<MPos> CellChanged = null;
+
+		public TerrainLighting(World world, TerrainLightingInfo info)
+		{
+			this.info = info;
+			map = world.Map;
+			globalTint = new float3(info.RedTint, info.GreenTint, info.BlueTint);
+
+			var cellSize = map.Grid.Type == MapGridType.RectangularIsometric ? 1448 : 1024;
+			partitionedLightSources = new SpatiallyPartitioned<LightSource>(
+				(map.MapSize.X + 1) * cellSize,
+				(map.MapSize.Y + 1) * cellSize,
+				info.BinSize * cellSize);
+		}
+
+		Rectangle Bounds(LightSource source)
+		{
+			var c = source.Pos;
+			var r = source.Range.Length;
+			return new Rectangle(c.X - r, c.Y - r, 2 * r, 2 * r);
+		}
+
+		public int AddLightSource(WPos pos, WDist range, float intensity, float3 tint)
+		{
+			var token = nextLightSourceToken++;
+			var source = new LightSource(pos, map.CellContaining(pos), range, intensity, tint);
+			var bounds = Bounds(source);
+			lightSources.Add(token, source);
+			partitionedLightSources.Add(source, bounds);
+
+			if (CellChanged != null)
+				foreach (var c in map.FindTilesInCircle(source.Cell, (source.Range.Length + 1023) / 1024))
+					CellChanged(c.ToMPos(map));
+
+			return token;
+		}
+
+		public void RemoveLightSource(int token)
+		{
+			LightSource source;
+			if (!lightSources.TryGetValue(token, out source))
+				return;
+
+			lightSources.Remove(token);
+			partitionedLightSources.Remove(source);
+			if (CellChanged != null)
+				foreach (var c in map.FindTilesInCircle(source.Cell, (source.Range.Length + 1023) / 1024))
+					CellChanged(c.ToMPos(map));
+		}
+
+		float3 ITerrainLighting.TintAt(WPos pos)
+		{
+			using (new PerfSample("terrain_lighting"))
+			{
+				var uv = map.CellContaining(pos).ToMPos(map);
+				if (!map.Height.Contains(uv))
+					return new float3(1, 1, 1);
```

Comment:
This can now use `float3.Ones` I suppose.

Reply:
I made it return `globalTint` instead.